DevCycle.SDK.Server.Local/Api/EventQueue.cs

```diff
@@ -174,91 +273,39 @@ public virtual void QueueAggregateEvent(DVCPopulatedUser user, Event @event, Buc
             eventCopy.Date = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
             eventCopy.Value = 1;
 
-            if (!userForAggregation.ContainsKey(user.UserId))
-            {
-                userForAggregation.Add(user.UserId, user);   
-            }
-            else
-            {
-                userForAggregation[user.UserId] = user;
-            }
-            
             var requestEvent = new DVCRequestEvent(
                 eventCopy, 
                 user.UserId, 
                 config == null ? new Dictionary<string, string>() : config.FeatureVariationMap
             );
             
-            AddAggregateEvent(requestEvent, user);
+            var userAndFeatureVars = new UserAndFeatureVars(user, requestEvent.FeatureVars);
 
+            aggregateEventQueueMutex.WaitOne();
+            aggregateEvents.AddEvent(userAndFeatureVars, requestEvent);
+            aggregateEventQueueMutex.ReleaseMutex();
             ScheduleFlushWithDelay();
         }
 
-        private void AddAggregateEvent(DVCRequestEvent requestEvent, DVCPopulatedUser user)
+        private Dictionary<DVCPopulatedUser, UserEventsBatchRecord> CombineUsersEventsToFlush()
         {
-            var aggregateUser = aggregateEvents.ContainsKey(user.UserId)
-                ? aggregateEvents[user.UserId]
-                : new Dictionary<string, Dictionary<string, DVCRequestEvent>>();
-
-            var aggregateEventType = aggregateUser.ContainsKey(requestEvent.Type)
-                ? aggregateUser[requestEvent.Type]
-                : AddAggregateEventType(user, aggregateUser, requestEvent);
-
-            if (aggregateEventType.ContainsKey(requestEvent.Target))
-            {
-                aggregateEventType[requestEvent.Target].Value += Decimal.One;
-            }
-            else
-            {
-                aggregateEventType.Add(requestEvent.Target, requestEvent);
-            }
-        }
-
-        private Dictionary<string, DVCRequestEvent> AddAggregateEventType(DVCPopulatedUser user,
-            Dictionary<string, Dictionary<string, DVCRequestEvent>> aggregateUser,
-            DVCRequestEvent requestEvent)
-        {
-            aggregateUser.Add(requestEvent.Type, new Dictionary<string, DVCRequestEvent>());
-            aggregateUser[requestEvent.Type].Add(requestEvent.Target, requestEvent);
-            aggregateEvents.Add(user.UserId, aggregateUser);
-
-            return aggregateUser[requestEvent.Type];
-        }
-
-        private Dictionary<string, UserEventsBatchRecord> CombineUsersEventsToFlush()
-        {
-            var userEventsBatchRecords = new Dictionary<string, UserEventsBatchRecord>();
+            var userEventsBatchRecords = aggregateEvents.GetEventBatches();
 
             foreach (var eventPayload in eventPayloadsToFlush)
             {
-                var userId = eventPayload.Key;
+                var user = eventPayload.Key;
```

Comment:
you could probably split this into a K/V foreach since you're already stripping them out.

Reply:
not sure what you mean by this

Reply:
not really relevant with the hashcode changes - the idea is to convert the keyvaluepair object into its raw key/values right in the for loop instead - but it breaks how other things would work